libretro-rs/src/core_macro.rs

```diff
@@ -4,6 +4,8 @@ macro_rules! libretro_core {
     #[doc(hidden)]
     mod __libretro_rs_gen {
       use super::*;
+      use core::ffi::*;
+      use core::ffi::c_char;
```

Comment:
You'd think that, but there's a second star import bringing in `c_char` so I need that to disambiguate it :) Figuring out where the second c_char is coming from is still on my to-do list...it's tough when it happens inside a macro.

Reply:
> You'd think that, but there's a second star import bringing in `c_char` so I need that to disambiguate it :) Figuring out where the second c_char is coming from is still on my to-do list...it's tough when it happens inside a macro.

The preference would be to then just import the things we need from `core::ffi`, since it would be confusing otherwise (case and point: I just got it wrong 😉).

> Re: Formatting, I've been relying on CLion for formatting, I'll be sure to use `rustfmt` moving forward.

👍🏼 if you don't have it already, I also recommend `editorconfig` as that'll keep some basic editor settings aligned project-wide. I assume CLion just supports it out of the box though.

> Re: The large amount of breakage, I'm aware I just upended a ton of the crate's API, but if you were planning on release a major version of it anyways, I think it makes sense to rip off the bandaid and do all the necessary changes to future-proof it now instead of waiting for someone to request a backwards-incompatible change because they can't get at some functionality they need.

Having slept on it, I think the best way to approach this would be to divvy up this PR into a few separate ones. A few concrete reasons for doing so:

1. This is "all or nothing", I can either accept all of it or reject all of it. Considering there are a few things I'd like to at least discuss in depth, my inclination as of now would be to reject it but I don't really want to do that since there's a lot of good here.
2. This is hard to consider all at once, and covers a lot of different areas.
3. I'd like to establish a workflow where issues are made to discuss changes first, then PRs follow to address the issues. I'm guilty of making PRs like this, so do as I say, not as I do 😜 I think this will be a better way to collaborate and save time if a particular change is too radical or if it needs to "soak" a while.

I would almost immediately accept the `.gitignore` changes, and the `libretro_core!` macro changes (with the note I just left about star imports in that file). The rest will need some more hashing out.

Reply:
That's fair. I still wanted to open the PR so you could see all the changes together and try them out.

Normally I prefer opening PRs for individual changes but I honestly didn't know how this was going to turn out going into it. I'm still fairly inexperienced when it comes to Rust API design and how traits and ownership will work out in practice.

I take it you want to start with the environment changes?

Reply:
> That's fair. I still wanted to open the PR so you could see all the changes together and try them out.

Open as many as you feel is right, and we can work through them 😄

> Normally I prefer opening PRs for individual changes but I honestly didn't know how this was going to turn out going into it. I'm still fairly inexperienced when it comes to Rust API design and how traits and ownership will work out in practice.

You're doing a great job! No need to apologize. I am very grateful for your contributions!

> I take it you want to start with the environment changes?

Again, whatever you feel is right. Just aim for more bite sized PRs and we'll be able to merge them more rapidly.

Thanks again for putting this together, it is very much appreciated!